Optimistic locking

dranikpg opened the pull request:
Hi! I've implemented optimistic locking as discussed [here](https://github.com/go-rel/rel/issues/100).

It's automatically enabled by including a `LockVersion int` field into the struct. 

It adds a `where lock_version = $` clause to select only valid records and uses `set lock_version = lock_version + 1` for updating locks.

### What I've changed

Mainly `repository.update()` and `repository.delete()`. They now include custom logic for applying locks.

Because using SQL expressions like _+=_ requires reloading the whole struct, I've added a `NoReload` flag to `Mutates` to surpress triggering reloads for lock updates. In case no reload is required, the locks are incremented manually.

### Tests

I've added simple tests for updating and deleting actual and stale records.

dranikpg commented:
@Fs02 The reltest setup somewhat confuses me 🤔 An older version of it is still under rel/reltest? How should I push a breaking API change (`Repository.Delete()`) so that rel and reltest stay in sync?

Fs02 commented:
@dranikpg ah right, let me remove it first

Fs02 commented:
Removed deprecated reltest: https://github.com/go-rel/rel/pull/266

dranikpg commented:
That still leaves me wondering how to update both rel and reltest as they depend on each other. The CI currently breaks because reltest still uses the old `Repository` interface. Updating reltest requires having an updated version of rel itself 🔁 

One solution is to specify dependencies with tags/branches during PR reviews and then change them before merging.

Fs02 commented:
sorry for the wait, I've decided to extract migrator package, now rel doesn't depends on reltest anymore
https://github.com/go-rel/rel/pull/267

dranikpg commented on document.go:
```diff
@@ -517,13 +520,23 @@ func extractBoolFlag(name string) DocumentFlag {
 }
 
+func extractIntFlag(name string) DocumentFlag {
+	if name == "lock_version" {
```
Field names like "created_at" and "updated_at" are also not constants, so I've also kept this a literal everywhere 🤷🏻

dranikpg commented on mutation.go:
```diff
@@ -125,9 +130,10 @@ const (
+	Type     ChangeOp
+	Field    string
+	Value    interface{}
+	NoReload Reload
```
Skip reload for queries that use SQL expressions which can be evaluated without querying

dranikpg commented on rel_test.go:
```diff
@@ -60,8 +60,17 @@ type Transaction struct {
 	AddressID int
-	Address   Address
-	Histories *[]History
+	Address   Address    `ref:"address_id" fk:"id"`
+	Histories *[]History `ref:"id" fk:"transaction_id"`
```
Auto field name deduction doesn't work with embedded structs 😞  I'll fix this

dranikpg commented on repository.go:
```diff
@@ -491,6 +491,18 @@ func (r repository) Update(ctx context.Context, record interface{}, mutators ...
 }
 
+func (r repository) lockVersion(doc Document, unscoped Unscoped) (int, bool) {
+	if unscoped {
+		return 0, false
+	}
```
See comments below

dranikpg commented on repository.go:
```diff
@@ -499,25 +511,8 @@ func (r repository) update(cw contextWrapper, doc *Document, mutation Mutation,
-			if err := r.find(cw, doc, query.UsePrimary()); err != nil {
-				return err
-			}
+		if err := r.applyMutates(cw, doc, mutation, filter); err != nil {
+			return err
```
To reduce function complexity

dranikpg commented on repository.go:
```diff
@@ -534,6 +529,43 @@ func (r repository) update(cw contextWrapper, doc *Document, mutation Mutation,
 
+func (r repository) applyMutates(cw contextWrapper, doc *Document, mutation Mutation, filter FilterQuery) error {
+	var (
+		baseQueries = []Querier{filter, mutation.Unscoped, mutation.Cascade}
+		queries     = baseQueries
```
We have to reload mutation queriers (like version_lock += 1) when reloading a document, so we store the base slice

dranikpg commented on repository.go:
```diff
@@ -534,6 +529,43 @@ func (r repository) update(cw contextWrapper, doc *Document, mutation Mutation,
+		baseQueries = []Querier{filter, mutation.Unscoped, mutation.Cascade}
+		queries     = baseQueries
+	)
+	if version, ok := r.lockVersion(*doc, mutation.Unscoped); ok {
+		Inc("lock_version").SkipReload().Apply(doc, &mutation)
+		queries = append(queries, LockVersion(version))
```
And add a query + mutation if the document has a version lock and the query is scoped

dranikpg commented on repository.go:
```diff
@@ -534,6 +529,43 @@ func (r repository) update(cw contextWrapper, doc *Document, mutation Mutation,
+
+	var (
+		pField string
+		query  = r.withDefaultScope(doc.data, Build(doc.Table(), queries...), false)
```
I first thought of reusing withDefaultScope for applying version_lock mutations

dranikpg commented on repository.go:
```diff
@@ -534,6 +529,43 @@ func (r repository) update(cw contextWrapper, doc *Document, mutation Mutation,
+		return NotFoundError{}
+	}
+
+	if mutation.Reload {
+		baseQuery := r.withDefaultScope(doc.data, Build(doc.Table(), baseQueries...), false)
+		if err := r.find(cw, doc, baseQuery.UsePrimary()); err != nil {
+			return err
+		}
+	} else if version, ok := r.lockVersion(*doc, mutation.Unscoped); ok {
+		doc.SetValue("lock_version", version+1)
+	}
+
```
But abstracting away version_lock logic doesn't allow us to simply modify the version lock afterwards. 

Otherwise we could extend mutates and mark some of them to be "not included" in reload queries and to be computable without a reload. This would, for example, allow handling other increments with a version lock without reloading the document.

dranikpg commented on repository.go:
```diff
@@ -758,37 +790,39 @@ func (r repository) MustUpdateAny(ctx context.Context, query Query, mutates ...M
-		cascade = Cascade(false)
+		cw       = fetchContext(ctx, r.rootAdapter)
+		doc      = NewDocument(record)
+		mutation = applyMutators(nil, false, false, mutators...)
```
Apply mutators without enabling cascading and "structset"

Fs02 commented on repository.go:
```diff
@@ -534,6 +529,43 @@ func (r repository) update(cw contextWrapper, doc *Document, mutation Mutation,
+			return err
+		}
+	} else if version, ok := r.lockVersion(*doc, mutation.Unscoped); ok {
+		doc.SetValue("lock_version", version+1)
```
I think this will return incorrect version if multiple process is updating the same record concurrently without using lock

to make things simple, I think what we can do is:
- if lock is **enabled**: automatically update using Set with new version value is computed from rel side
- if lock is **disabled**: don't automatically update version, warn user that lock version will be replaced

dranikpg replied:
Sorry for the long pause, had to take a break.

Why? If `SET version = version + 1 WHERE version = x` exectues successfully, then the version has to be x + 1. As far as I understand, we don't care about other fields. Mixing usage with and without locks is up to the user.

Fs02 replied:
> If SET version = version + 1 WHERE version = x exectues successfully, then the version has to be x + 1.

ah right this is true

but I can see the implementation can be simplified if we just do `SET version = newVersion WHERE version = previousVersion`, where we locally compute `newVersion = previousVersion+1` in memory

this way, we don't have to introduce NoReload, semantics that rel.Inc/rel.Dec/rel.Fragment always trigger reload will remain true.
(it always trigger reload, because there's no way to compute the updated record correctly in memory)

dranikpg replied:
Yeah, thats true.

They key problem here is that Set() calls are always applied to the document beforehand (in Mutate.Apply()), regardless of whether the record is found/any erros occured. This is the default behaviour for rel now. 
Should version locks be rolled back in case of unsuccessful updates?

Fs02 replied:
ah true,

> Should version locks be rolled back in case of unsuccessful updates?

I think yes

dranikpg commented:
I've removed all NoReload stuff and replaced the increment with a Set. A simple defer now rollbacks the version lock. 

Deepsource is failing because of some old issues.

Fs02 commented on repository_test.go:
```diff
@@ -3958,3 +3958,71 @@ func TestRepository_Transaction_runtimeError(t *testing.T) {
+	adapter.AssertExpectations(t)
+}
+
+func TestRepository_LockVersion_Delete(t *testing.T) {
```
can you add a test for LockVersion + SoftDelete

dranikpg replied:
Added. If I understand correctly, soft deletes don't update the documents fields?

Fs02 reviewed (approved):
LGTM, Thank you!!
